Xamarin.PropertyEditing.Mac/Controls/DateTimeEditorControl.cs

```diff
@@ -0,0 +1,83 @@
+﻿using System;
+using System.Collections;
+using AppKit;
+using Xamarin.PropertyEditing.Mac.Resources;
+using Xamarin.PropertyEditing.ViewModels;
+
+namespace Xamarin.PropertyEditing.Mac
+{
+	internal class DateTimeEditorControl : PropertyEditorControl<PropertyViewModel<DateTime>>
+	{
+		private const int HeightMargin = 1;
+
+		private readonly NSDatePicker editor;
+
+		public DateTimeEditorControl (IHostResourceProvider hostResources)
+			: base (hostResources)
+		{
+			this.editor = new NSDatePicker {
+				TranslatesAutoresizingMaskIntoConstraints = false,
+				Font = NSFont.FromFontName (DefaultFontName, DefaultFontSize),
+				DatePickerElements = NSDatePickerElementFlags.HourMinuteSecond | NSDatePickerElementFlags.YearMonthDateDay
+			};
+
+			// update the value on keypress
+			this.editor.Activated += Editor_Activated;
+
+			AddSubview (this.editor);
+
+			AddConstraints (new[] {
+				NSLayoutConstraint.Create (this.editor, NSLayoutAttribute.Top, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Top, 1f, 1f),
+				NSLayoutConstraint.Create (this.editor, NSLayoutAttribute.Left, NSLayoutRelation.Equal, this,  NSLayoutAttribute.Left, 1f, 0),
+				NSLayoutConstraint.Create (this.editor, NSLayoutAttribute.Height, NSLayoutRelation.Equal, 1f, DefaultControlHeight),
+			});
+		}
+
+		private void Editor_Activated (object sender, EventArgs e) =>
+			ViewModel.Value = this.editor.DateValue.ToDateTime ();
+
+		public override NSView FirstKeyView => this.editor;
+		public override NSView LastKeyView => this.editor;
+
+		public override nint GetHeight (EditorViewModel vm)
+		{
+			return DefaultControlHeight + HeightMargin;
+		}
+
+		protected override void UpdateValue ()
+		{
+			this.editor.DateValue = ViewModel.Value.ToNSDate ();
+		}
+
+		protected override void HandleErrorsChanged (object sender, System.ComponentModel.DataErrorsChangedEventArgs e)
+		{
+			UpdateErrorsDisplayed (ViewModel.GetErrors (ViewModel.Property.Name));
+		}
+
+		protected override void UpdateErrorsDisplayed (IEnumerable errors)
+		{
+			if (ViewModel.HasErrors) {
+				SetErrors (errors);
+			} else {
+				SetErrors (null);
+				SetEnabled ();
+			}
+		}
+
+		protected override void SetEnabled ()
+		{
+			this.editor.Enabled = ViewModel.Property.CanWrite;
+		}
+
+		protected override void UpdateAccessibilityValues ()
+		{
+			this.editor.AccessibilityTitle = string.Format (LocalizationResources.AccessibilityDateTime, ViewModel.Property.Name);
+		}
+
+		protected override void Dispose (bool disposing)
```

Comment:
Is this actually necessary?

Reply:
This was a long discussion related to memory allocation leaks in Xamarin.Mac over handlers. I prefer to invoke @Therzok which know better all the bit and surely can explain in detail the reason for this. But yes, we have a rule in MD to remove all the handling in Dispose.

Reply:
@ermau there's scenarios where ref-cycles do happen and only an event handler removal does fix it.

Even though Xamarin.Mac has weak event handling, Gtk# does not, so if we get through a toggleref to toggleref cycle between Gtk# widgets via this, we will end up leaking.

Just adds safety and the possibility to fix leaks via explicit disposal.

Reply:
Afaik, the eventhandler is stored via strong references, the target is stored via weak ones. So we might still encounter a refcycle via eventhandlers even in pure Xamarin.Mac.

I think there's an okay explanation here: https://github.com/mono/monodevelop/wiki/UI-code---Reference-counting-and-you#xamarinmac

tl;dr: Best of any world is actually having a static selector/eventhandler exported and using that, to avoid any kind of captures that would lead to refcycles.